Sources/QuickCheck/ArbitraryDecodable.swift

```diff
@@ -0,0 +1,218 @@
+
+private final class ArbitraryDecoder: Decoder {
+  var genState: GenState
+  var codingPath: [CodingKey] = []
+  let userInfo: [CodingUserInfoKey: Any] = [:]
+
+  init(seed: Seed, size: Int) {
+    self.genState = (seed, size)
+  }
+
+  func run<T>(_ gen: Gen<T>) -> T {
+    let (result, genState) = gen.run(self.genState)
+    self.genState = genState
+    return result
+  }
+
+  func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> where Key: CodingKey {
+    return .init(KeyedContainer(decoder: self))
+  }
+
+  func unkeyedContainer() throws -> UnkeyedDecodingContainer {
+    return UnkeyedContainer(decoder: self)
+  }
+
+  func singleValueContainer() throws -> SingleValueDecodingContainer {
+    return SingleValueContainer(decoder: self)
+  }
+
+  struct KeyedContainer<Key: CodingKey>: KeyedDecodingContainerProtocol {
+    let allKeys: [Key]
+    let codingPath: [CodingKey]
+    let decoder: ArbitraryDecoder
+
+    init(decoder: ArbitraryDecoder) {
+      let genKeys = arrayOf(genInt.map(Key.init(intValue:))) |> scale { $0 * 2 }
+
+      self.allKeys = decoder.run(genKeys).flatMap { $0 }
+      self.codingPath = decoder.codingPath
+      self.decoder = decoder
+    }
+
+    func contains(_ key: Key) -> Bool {
+      return true
+    }
+
+    func decodeNil(forKey key: Key) throws -> Bool {
+      return self.decoder.run(uniform.map { $0 >= 0.75 })
+    }
+
+    func decode<T>(_ type: T.Type, forKey key: Key) throws -> T where T: Decodable {
+      self.decoder.codingPath.append(key)
+      defer { self.decoder.codingPath.removeLast() }
+      return try T(from: self.decoder)
+    }
+
+    func nestedContainer<NestedKey>(keyedBy type: NestedKey.Type, forKey key: Key) throws
+      -> KeyedDecodingContainer<NestedKey> where NestedKey: CodingKey {
+        self.decoder.codingPath.append(key)
+        defer { self.decoder.codingPath.removeLast() }
+        return .init(KeyedContainer<NestedKey>(decoder: self.decoder))
+    }
+
+    func nestedUnkeyedContainer(forKey key: Key) throws -> UnkeyedDecodingContainer {
+      self.decoder.codingPath.append(key)
+      defer { self.decoder.codingPath.removeLast() }
+      return UnkeyedContainer(decoder: self.decoder)
+    }
+
+    func superDecoder() throws -> Decoder {
+      return self.decoder
+    }
+
+    func superDecoder(forKey key: Key) throws -> Decoder {
+      self.decoder.codingPath.append(key)
+      defer { self.decoder.codingPath.removeLast() }
+      return self.decoder
+    }
+  }
+
+  struct UnkeyedContainer: UnkeyedDecodingContainer {
+    let codingPath: [CodingKey]
+    let count: Int?
+    var currentIndex = 0
+    let decoder: ArbitraryDecoder
+    var isAtEnd: Bool { return currentIndex == count }
+
+    init(decoder: ArbitraryDecoder) {
+      self.codingPath = decoder.codingPath
+      print(decoder.genState.size)
+      self.count = decoder.run(choose(0..<decoder.genState.size) |> suchThat <| { $0 % 2 == 0 })
```

Comment:
Decoders don't decode `Array` as a type, so I'm using the logic behind the `arrayOf` generator to build up a random count. The `suchThat <| { $0 % 2 == 0 }` is a workaround for collections like `Set` which require even keys—hoping there's a better way.